### PR TITLE
P1949R7 C++ Identifier Syntax using Unicode Standard Annex 31

### LINKED ISSUE #4634: [2021-06 CWG Motion 4] P1949R7 C++ Identifier Syntax using Unicode Standard Annex 31
[P1949R7](https://wg21.link/p1949r7)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes  #4634
Fixes cplusplus/papers#688

### CODE REVIEW

File: source/back.tex [hunk 1]
```diff
@@ -23,6 +23,12 @@ \chapter{Bibliography}
   \doccite{Unicode Text Segmentation} [online].
   Edited by Mark Davis. Revision 35; issued for Unicode 12.0.0. 2019-02-15 [viewed 2020-02-23].
   Available from: \url{http://www.unicode.org/reports/tr29/tr29-35.html}
+\item
+  The Unicode Consortium. Unicode Standard Annex, UAX \#31,
+  \doccite{Unicode Identifier and Pattern Syntax} [online].
+  Edited by Mark Davis. Revision 33; issued for Unicode 13.0.0.
+  2020-02-13 [viewed 2021-06-08].
+  Available from: \url{https://www.unicode.org/reports/tr31/tr31-33.html}
```
Comment (JohelEGP) [inline source/back.tex:31]: This whole list item is different that what's at http://wg21.link/P1949R7#wording-for-uax-31-identifiers. It mentions UAX 44 (as is done below), but the wording above mentions UAX 31.

File: source/uax31.tex [hunk 2]
```diff
@@ -0,0 +1,131 @@
+%!TEX root = std.tex
+\infannex{uaxid}{Conformance with UAX \#31}
+
+\rSec1[uaxid.general]{General}
+
+\pnum
+This subclause describes the choices made in application of
```
Comment (JohelEGP) [inline source/uax31.tex:7]:
```suggestion
This Annex describes the choices made in application of
```
Like http://wg21.link/depr.general.

File: source/uax31.tex [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+UAX \#31 (``Unicode Identifier and Pattern Syntax'')
+to C++ in terms of the requirements from UAX \#31 and
```
Comment (JohelEGP) [inline source/uax31.tex:9]:
```suggestion
to \Cpp{} in terms of the requirements from UAX \#31 and
```

File: source/uax31.tex [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+how they do or do not apply to C++.
```
Comment (JohelEGP) [inline source/uax31.tex:10]:
```suggestion
how they do or do not apply to \Cpp{}.
```

File: source/uax31.tex [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+In terms of UAX \#31,
+C++ conforms by meeting the requirements
```
Comment (JohelEGP) [inline source/uax31.tex:12]:
```suggestion
\Cpp{} conforms by meeting the requirements
```

File: source/uax31.tex [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+R1 ``Default Identifiers'' and
+R4 ``Equivalent Normalized Identifiers''.
+The other requirements, also listed below,
+are either alternatives not taken or do not apply to C++.
```
Comment (JohelEGP) [inline source/uax31.tex:16]:
```suggestion
are either alternatives not taken or do not apply to \Cpp{}.
```

File: source/uax31.tex [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+\rSec1[uaxid.default]{R1 Default identifiers}
+
+\pnum
+UAX \#31 specifies a default syntax for identifier
```
Comment (JohelEGP) [inline source/uax31.tex:21]:
```suggestion
UAX \#31 specifies a default syntax for identifiers
```

File: source/uax31.tex [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+based on properties from the Unicode Character Database, UAX \#44.
+The general syntax is
+\begin{codeblock}
+<Identifier> := <Start> <Continue>* (<Medial> <Continue>+)*
+\end{codeblock}
+where \tcode{<Start>} has the XID_Start property,
+\tcode{<Continue>} has the XID_Continue property, and
+\tcode{<Medial>} is a list of characters permitted between continue characters.
+For C++ we add the character U+005F, LOW LINE, or \tcode{_},
```
Comment (JohelEGP) [inline source/uax31.tex:30]:
```suggestion
For \Cpp{} we add the character U+005F, LOW LINE, or \tcode{_},
```

File: source/uax31.tex [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+<Continue> := <Start> + XID_Continue
+\end{codeblock}
+
+\pnum
+This is described in the C++ grammar in \ref{lex.name},
+where \grammarterm{identifier} is formed from
+\grammarterm{identifier-start} or
+\grammarterm{identifier} followed by \grammarterm{identifier-continue}.
+
+\rSec1[uaxid.rfmt]{R1a Restricted format characters}
```
Comment (JohelEGP) [inline source/uax31.tex:47]:
```suggestion
\rSec2[uaxid.rfmt]{R1a Restricted format characters}
```

File: source/uax31.tex [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+\pnum
+If an implementation of UAX \#31 wishes to allow format characters
+such as ZERO WIDTH JOINER or ZERO WIDTH NON-JOINER
+it must define a profile allowing them, or
+describe precisely which combinations are permitted.
+
+\pnum
+C++ does not allow format characters in identifiers, so this does not apply.
```
Comment (JohelEGP) [inline source/uax31.tex:56]:
```suggestion
\Cpp{} does not allow format characters in identifiers, so this does not apply.
```

File: source/uax31.tex [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+\rSec1[uaxid.stable]{R1b Stable identifiers}
```
Comment (JohelEGP) [inline source/uax31.tex:58]:
```suggestion
\rSec2[uaxid.stable]{R1b Stable identifiers}
```

File: source/uax31.tex [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+\pnum
+An implementation of UAX \#31 may choose to guarantee
+that identifiers are stable across versions of the Unicode Standard.
+Once a string qualifies as an identifier it does so in all future versions.
+
+\pnum
+C++ does not make this guarantee,
```
Comment (JohelEGP) [inline source/uax31.tex:66]:
```suggestion
\Cpp{} does not make this guarantee,
```

File: source/uax31.tex [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+\rSec1[uaxid.immutable]{R2 Immutable identifiers}
+
+\pnum
+An implementation may choose to guarantee that
+the set of identifiers will never change
+by fixing the set of code points allowed in identifiers forever.
+
+\pnum
+C++ does not choose to make this guarantee.
```
Comment (JohelEGP) [inline source/uax31.tex:78]:
```suggestion
\Cpp{} does not choose to make this guarantee.
```

File: source/uax31.tex [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+\pnum
+UAX \#31 describes how languages that use or interpret patterns of characters,
+such as regular expressions or number formats,
+may describe that syntax with Unicode properties.
+
+\pnum
+C++ does not do this as part of the language,
+deferring to library components for such usage of patterns.
+This requirement does not apply to C++.
```
Comment (JohelEGP) [inline source/uax31.tex:93]:
```suggestion
\Cpp{} does not do this as part of the language,
deferring to library components for such usage of patterns.
This requirement does not apply to \Cpp{}.
```

File: source/uax31.tex [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+\rSec1[uaxid.eqn]{R4 Equivalent normalized identifiers}
+
+\pnum
+UAX \#31 requires that implementations describe
+how identifiers are compared and considered equivalent.
+
+\pnum
+C++ requires that identifiers be in Normalization Form C and
```
Comment (JohelEGP) [inline source/uax31.tex:102]:
```suggestion
\Cpp{} requires that identifiers be in Normalization Form C and
```

File: source/uax31.tex [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+therefore identifiers that compare the same under NFC are equivalent.
+This is described in \ref{lex.name}.
+
+\rSec1[uaxid.eqci]{R5 Equivalent case-insensitive identifiers}
+
+\pnum
+C++ considers case to be significant in identifier comparison, and
+does not do any case folding.
+This requirement does not apply to C++.
```
Comment (JohelEGP) [inline source/uax31.tex:111]:
```suggestion
\Cpp{} considers case to be significant in identifier comparison, and
does not do any case folding.
This requirement does not apply to \Cpp{}.
```

File: source/uax31.tex [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+\rSec1[uaxid.filter]{R6 Filtered normalized identifiers}
+
+\pnum
+If any characters are excluded from normalization,
+UAX \#31 requires a precise specification of those exclusions.
+
+\pnum
+C++ does not make any such exclusions.
```
Comment (JohelEGP) [inline source/uax31.tex:120]:
```suggestion
\Cpp{} does not make any such exclusions.
```

File: source/uax31.tex [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+
+\rSec1[uaxid.filterci]{R7 Filtered case-insensitive identifiers}
+
+\pnum
+C++ identifiers are case sensitive, and
```
Comment (JohelEGP) [inline source/uax31.tex:125]:
```suggestion
\Cpp{} identifiers are case sensitive, and
```

File: source/uax31.tex [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+therefore this requirement does not apply.
+
+\rSec1[uaxid.hashtag]{R8 Hashtag identifiers}
+
+\pnum
+There are no hashtags in C++, so this requirement does not apply.
```
Comment (JohelEGP) [inline source/uax31.tex:131]:
```suggestion
There are no hashtags in \Cpp{}, so this requirement does not apply.
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/uax31.tex [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+the \tcode{<Continue>} characters are unmodified.
+In the grammar used in UAX \#31, this is
+\begin{codeblock}
+<Identifier> := <Start> <Continue>*
+<Start> := XID_Start + U+005F
+<Continue> := <Start> + XID_Continue
+\end{codeblock}
+
+\pnum
+This is described in the C++ grammar in \ref{lex.name},
```
Comment (JohelEGP) [inline source/uax31.tex:42]:
```suggestion
This is described in the \Cpp{} grammar in \ref{lex.name},
```

**Comment by jensmaurer** [comment]
@JohelEGP Thanks, all applied plus addressed the hanging paragraph.

**Pull request merged**